File: src/main/java/com/hileco/cortex/instructions/io/SAVE.java
package com.hileco.cortex.instructions.io;

import com.hileco.cortex.vm.VirtualMachine;
import com.hileco.cortex.vm.ProgramContext;
import com.hileco.cortex.vm.ProgramStoreZone;
import com.hileco.cortex.vm.layer.LayeredBytes;
import com.hileco.cortex.instructions.ProgramException;
import com.hileco.cortex.instructions.StackParameter;
import lombok.EqualsAndHashCode;

import java.math.BigInteger;
import java.util.List;

import static com.hileco.cortex.instructions.ProgramException.Reason.STACK_TOO_FEW_ELEMENTS;

@EqualsAndHashCode(callSuper = true)
public class SAVE extends IoInstruction {

    public static final StackParameter ADDRESS = new StackParameter("address", 0);
    public static final StackParameter BYTES = new StackParameter("bytes", 1);

    public SAVE(ProgramStoreZone programStoreZone) {
        super(programStoreZone);
    }

    @Override
    public void execute(VirtualMachine process, ProgramContext program) throws ProgramException {
        var stack = program.getStack();
        if (stack.size() < 2) {
            throw new ProgramException(program, STACK_TOO_FEW_ELEMENTS);
        }
        var addressBytes = stack.pop();
        var address = new BigInteger(addressBytes);
        var bytes = stack.pop();
        LayeredBytes layeredBytes;

        var programStoreZone = this.getProgramStoreZone();
        switch (programStoreZone) {
            case MEMORY:
                layeredBytes = program.getMemory();
                break;
            case DISK:
                layeredBytes = program.getProgram().getStorage();
                break;
            default:
                throw new IllegalArgumentException(String.format("Unsupported ProgramStoreZone: %s", programStoreZone));
        }
        layeredBytes.write(address.intValue(), bytes);
    }

    @Override
    public List<Integer> getStackAdds() {
        return List.of();
    }

    @Override
    public List<StackParameter> getStackParameters() {
        return List.of(ADDRESS, BYTES);
    }
}

File: src/main/java/com/hileco/cortex/instructions/io/SAVE.kt
package com.hileco.cortex.instructions.io

import com.hileco.cortex.constraints.expressions.Expression
import com.hileco.cortex.instructions.ProgramException
import com.hileco.cortex.instructions.ProgramException.Reason.STACK_UNDERFLOW
import com.hileco.cortex.instructions.ProgramException.Reason.STORAGE_ACCESS_OUT_OF_BOUNDS
import com.hileco.cortex.instructions.StackParameter
import com.hileco.cortex.vm.ProgramStoreZone
import com.hileco.cortex.vm.concrete.ProgramContext
import com.hileco.cortex.vm.concrete.VirtualMachine
import com.hileco.cortex.vm.layer.LayeredBytes
import com.hileco.cortex.vm.layer.LayeredMap
import com.hileco.cortex.vm.symbolic.SymbolicProgramContext
import com.hileco.cortex.vm.symbolic.SymbolicVirtualMachine
import java.math.BigInteger

class SAVE(programStoreZone: ProgramStoreZone) : IoInstruction(programStoreZone) {
    override val stackParameters: List<StackParameter>
        get() = listOf(ADDRESS, BYTES)

    override fun execute(virtualMachine: VirtualMachine, programContext: ProgramContext) {
        if (programContext.stack.size() < 2) {
            throw ProgramException(STACK_UNDERFLOW)
        }
        val addressBytes = programContext.stack.pop()
        val address = BigInteger(addressBytes)
        val storage: LayeredBytes = when (programStoreZone) {
            ProgramStoreZone.MEMORY -> programContext.memory
            ProgramStoreZone.DISK -> programContext.program.storage
            ProgramStoreZone.CALL_DATA -> throw IllegalArgumentException("Unsupported ProgramStoreZone: $programStoreZone")
        }
        val bytes = programContext.stack.pop()
        val alignmentOffset = LOAD.SIZE - bytes.size
        if (address.toInt() * LOAD.SIZE + alignmentOffset + bytes.size > storage.size) {
            throw ProgramException(STORAGE_ACCESS_OUT_OF_BOUNDS)
        }
        storage.write(address.toInt() * LOAD.SIZE + alignmentOffset, bytes)
    }

    override fun execute(virtualMachine: SymbolicVirtualMachine, programContext: SymbolicProgramContext) {
        if (programContext.stack.size() < 2) {
            throw ProgramException(STACK_UNDERFLOW)
        }
        val addressExpression = programContext.stack.pop() as? Expression.Value
                ?: throw UnsupportedOperationException("Non-concrete address loading is not supported for symbolic execution")
        val storage: LayeredMap<BigInteger, Expression> = when (programStoreZone) {
            ProgramStoreZone.MEMORY -> programContext.memory
            ProgramStoreZone.DISK -> programContext.program.storage
            ProgramStoreZone.CALL_DATA -> throw IllegalArgumentException("Unsupported ProgramStoreZone: $programStoreZone")
        }
        val valueExpression = programContext.stack.pop()
        storage[addressExpression.constant.toBigInteger()] = valueExpression
    }

    companion object {
        val ADDRESS = StackParameter("address", 0)
        val BYTES = StackParameter("bytes", 1)
    }
}
